File: helix/src/main/java/glitch/helix/service/request/VideoRequest.java
package glitch.helix.service.request;

import glitch.api.http.HttpClient;
import glitch.api.http.Routes;
import glitch.api.objects.enums.VideoType;
import glitch.helix.object.enums.Period;
import glitch.helix.object.enums.VideoSort;
import glitch.helix.object.json.Game;
import glitch.helix.object.json.User;
import glitch.helix.object.json.Video;
import glitch.helix.object.json.Videos;
import glitch.service.AbstractRequest;
import java.util.Locale;
import reactor.core.publisher.Mono;

public class VideoRequest extends AbstractRequest<Video, Videos> {

    private String after;
    private String before;
    private Integer first;
    private Locale language;
    private Period period;
    private VideoSort sort;
    private VideoType type;

    public VideoRequest(HttpClient httpClient, User user) {
        super(httpClient, Routes.get("/videos").newRequest().queryParam("user_id", user.getId()));
    }

    public VideoRequest(HttpClient httpClient, Game game) {
        super(httpClient, Routes.get("/videos").newRequest().queryParam("user_id", game.getId()));
    }

    public VideoRequest setAfter(String after) {
        this.after = after;
        this.before = null;
        return this;
    }

    public VideoRequest setBefore(String before) {
        this.before = before;
        this.after = null;
        return this;
    }

    public VideoRequest setFirst(Integer first) {
        this.first = first;
        return this;
    }

    public VideoRequest setLanguage(Locale language) {
        this.language = language;
        return this;
    }

    public VideoRequest setPeriod(Period period) {
        this.period = period;
        return this;
    }

    public VideoRequest setSort(VideoSort sort) {
        this.sort = sort;
        return this;
    }

    public VideoRequest setType(VideoType type) {
        this.type = type;
        return this;
    }

    @Override
    public Mono<Videos> get() {
        if (before != null) {
            request.queryParam("before", before);
        }

        if (after != null) {
            request.queryParam("after", after);
        }

        if (first != null && (first > 0 && first <= 100)) {
            request.queryParam("first", first);
        }

        if (language != null) {
            request.queryParam("language", language.toLanguageTag());
        }

        if (period != null) {
            request.queryParam("period", period.name().toLowerCase());
        }

        if (sort != null) {
            request.queryParam("sort", sort.name().toLowerCase());
        }

        if (type != null) {
            request.queryParam("type", type.name().toLowerCase());
        }

        return httpClient.exchangeAs(request, Videos.class);
    }
}

File: helix/src/main/kotlin/io/glitchlib/api/service/request/VideoRequest.kt
package io.glitchlib.api.service.request

import io.glitchlib.api.model.Game
import io.glitchlib.api.model.Period
import io.glitchlib.api.model.User
import io.glitchlib.api.model.VideoSort
import io.glitchlib.internal.http.AbstractRequest
import io.glitchlib.internal.http.HttpRequest
import io.glitchlib.model.VideoType
import java.util.*

class VideoRequest : AbstractRequest {
    internal constructor(user: User) {
        this.userId = user.id
        this.gameId = null
    }

    internal constructor(game: Game) {
        this.gameId = game.id
        this.userId = null
    }

    private val userId: Long?
    private val gameId: Long?

    private var limit: Int? = null
    private var after: String? = null
    private var before: String? = null

    private var language: Locale? = null
    private var period: Period? = null
    private var sort: VideoSort? = null
    private var type: VideoType? = null

    fun limit(limit: Int) = apply {
        this.limit = limit
    }

    fun after(after: String) = apply {
        this.after = after
    }

    fun before(before: String) = apply {
        this.before = before
    }

    fun language(language: Locale) = apply {
        this.language = language
    }

    fun period(period: Period) = apply {
        this.period = period
    }

    fun sort(sort: VideoSort) = apply {
        this.sort = sort
    }

    fun type(type: VideoType) = apply {
        this.type = type
    }

    override fun invoke(): HttpRequest.() -> Unit = {
        if (userId != null) addQueryParameters("user_id", userId.toString())
        if (gameId != null) addQueryParameters("game_id", gameId.toString())
        if (limit != null && limit!! in 1..100) addQueryParameters("first", limit!!.toString())
        if (after != null) addQueryParameters("after", after!!)
        if (before != null) addQueryParameters("before", before!!)
        if (language != null) addQueryParameters("language", language!!.language)
        if (period != null) addQueryParameters("period", period!!.name.toLowerCase())
        if (sort != null) addQueryParameters("sort", sort!!.name.toLowerCase())
        if (type != null) addQueryParameters("type", type!!.name.toLowerCase())
    }
}